flang/lib/Lower/Bridge.cpp

```diff
@@ -447,29 +459,131 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+      blockList.push_back(blockOfLabel(eval, std::get<1>(stmt.t)));
+      attrList.push_back(fir::LowerBoundAttr::get(context));
+      valueList.push_back(builder->createIntegerConstant(exprType, 1));
+      blockList.push_back(blockOfLabel(eval, std::get<3>(stmt.t)));
+      attrList.push_back(mlir::UnitAttr::get(context)); // 0 is the "default"
+      blockList.push_back(blockOfLabel(eval, std::get<2>(stmt.t)));
+      builder->create<fir::SelectCaseOp>(toLocation(), expr, attrList,
+                                         valueList, blockList);
+      return;
+    }
+    // Arithmetic expression has Real type.  Generate
+    //   sum = expr + expr  [ raise an exception if expr is a NaN ]
+    //   if (sum < 0.0) goto L1 else if (sum > 0.0) goto L3 else goto L2
+    assert(eval.localBlocks.size() == 1 && "missing arithmetic if block");
+    mlir::Value sum = builder->create<fir::AddfOp>(toLocation(), expr, expr);
+    mlir::Value zero = builder->create<mlir::ConstantOp>(
+        toLocation(), exprType, builder->getFloatAttr(exprType, 0.0));
+    mlir::Value cond1 = builder->create<mlir::CmpFOp>(
+        toLocation(), mlir::CmpFPredicate::OLT, sum, zero);
+    genFIRConditionalBranch(cond1, blockOfLabel(eval, std::get<1>(stmt.t)),
+                            eval.localBlocks[0]);
+    startBlock(eval.localBlocks[0]);
+    mlir::Value cond2 = builder->create<mlir::CmpFOp>(
+        toLocation(), mlir::CmpFPredicate::OGT, sum, zero);
+    genFIRConditionalBranch(cond2, blockOfLabel(eval, std::get<3>(stmt.t)),
+                            blockOfLabel(eval, std::get<2>(stmt.t)));
+  }
+
+  void genFIR(Fortran::lower::pft::Evaluation &eval,
+              const Fortran::parser::AssignedGotoStmt &stmt) {
+    // Program requirement 1990 8.2.4 -
+    //
+    //   At the time of execution of an assigned GOTO statement, the integer
+    //   variable must be defined with the value of a statement label of a
+    //   branch target statement that appears in the same scoping unit.
+    //   Note that the variable may be defined with a statement label value
+    //   only by an ASSIGN statement in the same scoping unit as the assigned
+    //   GOTO statement.
+
+    const auto &symbolLabelMap =
+        eval.getOwningProcedure()->assignSymbolLabelMap;
+    const auto &symbol = *std::get<Fortran::parser::Name>(stmt.t).symbol;
+    auto variable = localSymbols.lookupSymbol(symbol);
+    if (!variable)
+      variable = createTemporary(toLocation(), symbol);
+    auto selectExpr = builder->create<fir::LoadOp>(toLocation(), variable);
+    auto iter = symbolLabelMap.find(symbol);
+    if (iter == symbolLabelMap.end()) {
+      // This "assert" will fail for a nonconforming program unit that does not
+      // have any ASSIGN statements.  The front end should check for this.
+      // If asserts are inactive, the assigned GOTO statement will be a nop.
+      llvm_unreachable("no assigned goto targets");
```

Comment:
Peter is right that the front end can't catch **all** program requirement errors, but the front end can and **should** catch obvious, control-flow independent errors.  The code here effectively does that after the fact.  The front end should do the same and generate a normal, user friendly error message.  Without that front end semantics checking, this code must do something.  For now, if someone hits this "assert", the intent is for this to be a reminder to fix the front end.  If "asserts" are disabled, the goto will act as a nop continue statement, which seems like a reasonable response.  Since this is some of the most unimportant obsolete functionality in the compiler, there's not much motivation to ask for changes now (or look at making changes myself, which is another possibility).

Reply:
In the middle-end, we absolutely should use `llvm_unreachable` when we get into a situation that there is no alternative other than "trying to go forward with bad/broken/garbage input". That is much better than producing garbage output because we just drop something on the floor.